### PR TITLE
adds timeouts to ODIS signer

### PR DESCRIPTION
**Pull request by alecps**

### CODE REVIEW

File: packages/phone-number-privacy/signer/src/server.ts [hunk 1]
```diff
@@ -52,14 +53,28 @@ export function createServer() {
   }
 }
 
-async function callAndMeasureLatency(
+async function callWithTimeout(
```
Comment (codyborn) [inline packages/phone-number-privacy/signer/src/server.ts]: I think we can set this at the Express level in index.ts
```
server.listen(port, () => {
    logger.info(`Server is listening on port ${port}`)
  }).setTimeout(1000)
  ```
  https://nodejs.org/dist/latest-v6.x/docs/api/http.html#http_server_settimeout_msecs_callback

Comment (nambrot) [reply packages/phone-number-privacy/signer/src/server.ts]: +1

Though it would be nice in either approach to add a prometheus metric to track how many requests time out

Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts]: Yea my concern is that if we do it at the express level the latency measurement will never wrap up and we'll be left without insight into how many timeouts we're seeing

Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts]: What we'll see with this implementation when signers timeout is signers reporting latency measurements that hit the ceiling set by the timeout parameter. We'll also see a prometheus counter tracking the number of timeouts.

**Comment by nambrot** [comment]
One thing that is interesting is that while we would timeout the request, express will not actually cancel the processing of it

https://github.com/expressjs/timeout

File: packages/phone-number-privacy/signer/src/server.ts [hunk 2]
```diff
@@ -52,14 +54,23 @@ export function createServer() {
...
   res: Response
 ) {
+  const childLogger: Logger = res.locals.logger
   const end = Histograms.responseLatency.labels(endpoint).startTimer()
-  await handler(req, res).finally(end)
+  const timeoutRes = Symbol()
+  await timeout(handler, [req, res], config.timeout, timeoutRes)
+    .catch((error: any) => {
+      if (error === timeoutRes) {
+        Counters.signatureRequestsWithoutSessionID.inc()
```
Comment (nambrot) [inline packages/phone-number-privacy/signer/src/server.ts]: Is this the counter you want to be incrementing?

File: packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts [hunk 3]
```diff
@@ -44,7 +44,7 @@ export async function handleGetBlindedMessagePartialSig(
   logger.info({ request: request.body }, 'Request received')
   if (!request.body.sessionID) {
     logger.debug({ request: request.body }, 'Request does not have sessionID')
-    Counters.signatureRequestsWithoutSessionID.inc()
+    Counters.timeouts.inc()
```
Comment (nambrot) [inline packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts]: Is this change intentional?

File: packages/sdk/base/src/async.ts [hunk 4]
```diff
@@ -139,3 +139,29 @@ export async function concurrentValuesMap<IN extends any, OUT extends any>(
     return output
   }, {})
 }
+
+export const timeout = <T extends any[], U>(
+  inFunction: InFunction<T, U>,
```
Comment (nambrot) [inline packages/sdk/base/src/async.ts]: Shouldn't this be a Promise?

File: packages/sdk/base/src/async.ts [hunk 5]
```diff
@@ -139,3 +139,29 @@ export async function concurrentValuesMap<IN extends any, OUT extends any>(
...
+  params: T,
+  timeoutMs: number,
+  timeoutRes: any,
+  logger: Logger | null = null
+) => {
+  let timer: number
+  return Promise.race([
+    inFunction(...params),
+    new Promise(() => {
```
Comment (nambrot) [inline packages/sdk/base/src/async.ts]: How does this promise ever resolve?

File: packages/phone-number-privacy/signer/package.json [hunk 6]
```diff
@@ -18,6 +18,7 @@
     "ssl:keygen": "./scripts/create-ssl-cert.sh"
   },
   "dependencies": {
+    "@celo/base": "1.1.1-dev",
```
Comment (nambrot) [inline packages/phone-number-privacy/signer/package.json]: Will this work with our dockerfile building since its not available?

File: packages/sdk/base/src/async.ts [hunk 7]
```diff
@@ -139,3 +139,29 @@ export async function concurrentValuesMap<IN extends any, OUT extends any>(
     return output
   }, {})
 }
+
+export const timeout = <T extends any[], U>(
```
Comment (nambrot) [inline packages/sdk/base/src/async.ts]: Maybe add some comment here on what the timeout behavior is. Right now in here it's not obvious that it would throw an error, but that's what the test seems to indicate

**Review by nambrot — CHANGES_REQUESTED** [review]

File: packages/phone-number-privacy/signer/src/server.ts (hunk 1 above)
Comment (codyborn) [reply packages/phone-number-privacy/signer/src/server.ts]: I think we should add this at the server level too as a last resort and set a higher limit

File: packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts (hunk 3 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/signing/get-partial-signature.ts]: Ah yea I mixed these two up

File: packages/phone-number-privacy/signer/src/server.ts (hunk 1 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/src/server.ts]: 👍

File: packages/sdk/base/src/async.ts (hunk 5 above)
Comment (alecps) [reply packages/sdk/base/src/async.ts]: setTimeout **rejects** (throws an error) to whatever you pass in as its 3rd argument when the timeout is reached. The Promise that wraps it also rejects to the same value, as does the Promise.race.

File: packages/sdk/base/src/async.ts (hunk 4 above)
Comment (alecps) [reply packages/sdk/base/src/async.ts]: It could be..I guess it felt cleaner to create the promise inside the timeout function. It's also what the other functions here are doing and the code pattern was easier to copy. Let me know if you feel strongly about this

File: packages/sdk/base/src/async.ts (hunk 7 above)
Comment (alecps) [reply packages/sdk/base/src/async.ts]: will do

File: packages/phone-number-privacy/signer/package.json (hunk 6 above)
Comment (alecps) [reply packages/phone-number-privacy/signer/package.json]: hmm good question. I think this will also come up in Cody's PR

Comment (alecps) [reply packages/phone-number-privacy/signer/package.json]: Yea I think you're right. One workaround would be to import it in through the common package (which we're already doing in Cody's combiner timeout PR). We'd need to repubiish the package though

Comment (alecps) [reply packages/phone-number-privacy/signer/package.json]: Just pushed this change
